### PR TITLE
Feat/flask restful error handling

### PR DESCRIPTION
**Pull request by TeddyAndersson**
## Feature description
This feature adds the functionlity to catch all errors raised in the code base and format them to http error responses with the correct status code.

## Solution description
The flask_restful Api class have a function built in to catch all errors that is raised within the application. This function (handle_errors) can also be overwritten to handle raised errors in the way we want to. So the solution was to simply extend the flas_restful Api class and override the handle_errors function. 

## What areas is affected by these changes?.
These changes currently only effect the application_status resource since that's the only resource that take advantage of this new funtionality. All endpoints however do have support for this behavior, but since we handle the errors in eact endpoints by try except blocks the error_handler never triggers.

## Testing the changes?
To test these changes you need to do the following:

1. Make sure your environment is setup according to the readme file
2. Checkout this branch
3. run commands: $ pipenv shell-> $ pipenv install -> $ flask run
4. Make requests to the route application/{hash_id}/status, with invalid hash_ids of any kind
5. This should result in a 400 Bad Request Error response.

Further more if you break the code in some way the error handler will catch this and return a 500 Internal server error.

## Was this feature tested in the following environments?
- [x] Your personal VADA environment.

### CODE REVIEW

**Review by dannil76 — CHANGES_REQUESTED** [review]
Put ExtendedFlaskRestfulApi in a separate file

File: app/libs/personal_number_helper.py [hunk 1]
```diff
@@ -1,21 +1,30 @@
...
 
-    if not len(hash_id) == 32:
-        raise Fault(message='Invalid hash_id', code=400)
+    decoded_hash_id_tuple = decode_hash_id(hash_id)
+
+    if len(decoded_hash_id_tuple) == 0:
+        raise ValueError(
+            f'Expected decoded_hash_id_tuple to have atleast one value (a,) got {decoded_hash_id_tuple} instead')
+
+    personal_number = str(int_tuple[0])
```
Comment (dannil76) [inline app/libs/personal_number_helper.py:26]: decoded_hash_id_tuple

File: app/routes/application_status.py [hunk 2]
```diff
@@ -1,34 +1,29 @@
...
     def get(self, hash_id):
-        try:
-            if not hash_id:
-                raise TypeError('hash_id should be type string')
+        hash_id_tuple = decode_hash_id(hash_id=hash_id)
 
-            personal_number = hash_to_personal_number(hash_id=hash_id)
+        if len(hash_id_tuple) == 0:
+            raise ValidationError(
+                message=f'The hashid {hash_id} in the url is not a valid hashid')
```
Comment (dannil76) [inline app/routes/application_status.py:17]: happy hacker :)

File: app/routes/application_status.py [hunk 3]
```diff
@@ -1,34 +1,29 @@
...
+from ..errors import ValidationError
 
 
 class ApplicationStatus(Resource):
 
     def get(self, hash_id):
-        try:
-            if not hash_id:
-                raise TypeError('hash_id should be type string')
+        hash_id_tuple = decode_hash_id(hash_id=hash_id)
```
Comment (dannil76) [inline app/routes/application_status.py:13]: decoded_hash_id

File: app/routes/application_status.py [hunk 4]
```diff
@@ -1,34 +1,29 @@
...
 
-            personal_number = hash_to_personal_number(hash_id=hash_id)
+        if len(hash_id_tuple) == 0:
+            raise ValidationError(
+                message=f'The hashid {hash_id} in the url is not a valid hashid')
 
-            viva_application_status = VivaApplicationStatus(
-                personal_number=personal_number)
+        personal_number = hash_id_tuple[0]
```
Comment (dannil76) [inline app/routes/application_status.py:19]: decode_hash_id

**Review by dannil76 — CHANGES_REQUESTED** [review]
Left a couple of comments on parts if you might please be able to ponder

File: app/routes/application_status.py (hunk 2 above)
Comment (KenthLjungHBG) [reply app/routes/application_status.py:17]: Agree with this! No need to print any unsanitized input 👍 .

File: app/__init__.py [hunk 5]
```diff
@@ -16,7 +63,7 @@ def create_app():
 
         from . import routes
 
-        api = Api(app)
+        api = ExtendedFlaskRestfulApi(app)
```
[the remark this answers is not in the document]
Comment (TeddyAndersson) [reply app/__init__.py:66]: I think this class tells you very clearly that it's an extension of the flask_restful Api. What this extension is can be found in the class. From my view ApiWithCustomErrorHandling is more confusing and FlaskRestfulApiWithCustomErrorHandling is way to long. so CustomFlaskRestfulApi or ExtendedFlaskRestfulApi feels like a good in between.

File: app/routes/application_status.py [hunk 6]
```diff
@@ -4,19 +4,19 @@
...
     def get(self, hash_id):
-        hash_id_tuple = decode_hash_id(hash_id=hash_id)
+        decoded_hash_id = decode_hash_id(hash_id=hash_id)
 
-        if len(hash_id_tuple) == 0:
-            raise ValidationError(
-                message=f'The hashid {hash_id} in the url is not a valid hashid')
+        if len(decoded_hash_id) == 0:
+            raise CustomValidationError(
+                message=f'The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).')
```
Comment (dannil76) [inline app/routes/application_status.py:17]: Man... Hope this never happens. The web version 2000 comes out before I have read the error message.

**Review by dannil76 — CHANGES_REQUESTED** [review]
A little modest comment only then I am satisfied

File: app/routes/application_status.py (hunk 6 above)
Comment (TeddyAndersson) [reply app/routes/application_status.py:17]: This is the default Bad Request Error description, so not my own words. But if you have any better idea, i'm open for suggestions.

**Comment by helsingborgsstad** [comment]
Task linked: [CU-gq827w Se till så att adaptern aldrig skickar 200 status när fel fångas](https://app.clickup.com/t/gq827w)

**Review by dannil76 — APPROVED** [review]

**Pull request merged**